src/crewai/crew.py

```diff
@@ -607,74 +577,122 @@ def _run_hierarchical_process(self) -> CrewOutput:
             )
             self.manager_agent = manager
 
-        task_outputs: List[TaskOutput] = []
-        futures: List[Tuple[Task, Future[TaskOutput]]] = []
+    def _execute_tasks(
+        self,
+        tasks: List[Task],
+        manager: Optional[BaseAgent] = None,
+        start_index: Optional[int] = 0,
+        was_replayed: bool = False,
+    ) -> CrewOutput:
+        """Executes tasks sequentially and returns the final output.
 
-        # TODO: IF USER OVERRIDE THE CONTEXT, PASS THAT
-        for task in self.tasks:
-            self._logger.log("debug", f"Working Agent: {manager.role}")
-            self._logger.log("info", f"Starting Task: {task.description}")
+        Args:
+            tasks (List[Task]): List of tasks to execute
+            manager (Optional[BaseAgent], optional): Manager agent to use for delegation. Defaults to None.
 
-            if self.output_log_file:
-                self._file_handler.log(
-                    agent=manager.role, task=task.description, status="started"
+        Returns:
+            CrewOutput: Final output of the crew
+        """
+        task_outputs: List[TaskOutput] = []
+        futures: List[Tuple[Task, Future[TaskOutput], int]] = []
+        last_sync_output: Optional[TaskOutput] = None
+
+        for task_index, task in enumerate(tasks):
+            if start_index is not None and task_index < start_index:
```

Comment:
Let's add some comments explaining that this code supports crew replays.

Reply:
Also, let's break this up into 2 for loops.

Right now, we are making this one for loop do too much: handle existing replayed tasks and then execute new tasks.

ideally, we should have the following. Screenshot below.

Please let me know if you have any questions.

Reply:
<img width="975" alt="Screenshot 2024-07-12 at 11 36 55 AM" src="https://github.com/user-attachments/assets/2f5d5b00-0b4f-4ca6-adc9-1bc525c8bb23">